### PR TITLE
Typecheck javascript

### PR DESCRIPTION
**Pull request by jelly**
Let's typecheck JavaScript! With the typecheck script we can also run it on JavaScript and filter out all errors we get and error by error remove them. As for example in this PR, spotted a silly typo :)

One issue is that running with `--javascript` and then without busts the cache, so maybe we just always want to run with `--javascript` as that should probably work filtering based on file and error list.

### CODE REVIEW

**Review by martinpitt — CHANGES_REQUESTED** [review]
This is a different vector/approach than the "relaxed" marker. I'm fine with supporting both and seeing whether a check-by-check or file-by-file approach works better -- and it's nice that we have a choice and can pick the best for the current situation.

But I'd like to see a strategy how we put that into CI. Right now this is entirely optional and dev-side only, and nothing prevents PRs from regressing JS files again. So can test/static-code call the script with `--javascript` a second time? How is that related to the cache issue you mentioned?

Note the [ruff error](https://github.com/cockpit-project/cockpit/actions/runs/12143443140/job/33860329160?pr=21377#step:2:475), which needs to be fixed for this to land.

Thanks!

**Comment by jelly** [comment]
> This is a different vector/approach than the "relaxed" marker. I'm fine with supporting both and seeing whether a check-by-check or file-by-file approach works better -- and it's nice that we have a choice and can pick the best for the current situation.

I'm not sure if I like adding markers in JavaScript files which are totally non-related to JavaScript. I want to experiment with removing errors one by one. And seeing how it goes.

> But I'd like to see a strategy how we put that into CI. Right now this is entirely optional and dev-side only, and nothing prevents PRs from regressing JS files again. So can test/static-code call the script with `--javascript` a second time? How is that related to the cache issue you mentioned?

Well that would be running with `--javascript`, but as I said if you run with or without javascript it seems to invalidate the TS cache so this will be slower. So we could always run on JS and simply have JS with custom filter rules.

> Note the [ruff error](https://github.com/cockpit-project/cockpit/actions/runs/12143443140/job/33860329160?pr=21377#step:2:475), which needs to be fixed for this to land.

That's fixed locally.

**Comment by martinpitt** [comment]
> I'm not sure if I like adding markers in JavaScript files which are totally non-related to JavaScript. I want to experiment with removing errors one by one. And seeing how it goes.

Yes, I agree, that's what I meant with my reply. :+1: 

> So we could always run on JS

So `--javascript` isn't an either/or, it will check .ts *and* .js? Why isn't that the default then? Or are there still too many errors in our .js files which aren't yet ignored by that long ignore list?

My point is: If we land this as-is, this is neither enforced by CI nor documented. If that's an experiment, that's ok, but I'd like to know how we get from here to "CI checks and gates on this". As otherwise we'll regress this all the time.

**Comment by jelly** [comment]
> > I'm not sure if I like adding markers in JavaScript files which are totally non-related to JavaScript. I want to experiment with removing errors one by one. And seeing how it goes.
> 
> Yes, I agree, that's what I meant with my reply. 👍
> 
> > So we could always run on JS
> 
> So `--javascript` isn't an either/or, it will check .ts _and_ .js? Why isn't that the default then? Or are there still too many errors in our .js files which aren't yet ignored by that long ignore list?

I should have marked this as draft :) I started with optional `--javascript` mode and seeing how far I could come. Turns out pretty far *cough* fixing one small error :)

> My point is: If we land this as-is, this is neither enforced by CI nor documented. If that's an experiment, that's ok, but I'd like to know how we get from here to "CI checks and gates on this". As otherwise we'll regress this all the time.

will do!

**Comment by jelly** [comment]
@martinpitt GitHub hides your comment but I've adjusted the commit message :)

**Review by martinpitt — APPROVED** [review]
Thanks! This works for me. @mvollmer can you please check, too?

File: pkg/storaged/iscsi/session.jsx [hunk 1]
```diff
@@ -47,7 +47,7 @@ export function make_iscsi_session_page(parent, session) {
         page_location: ["iscsi", session.data.target_name],
         page_name: session.data.target_name,
         page_icon: NetworkIcon,
-        page_categroy: PAGE_CATEGORY_NETWORK,
+        page_category: PAGE_CATEGORY_NETWORK,
```
Comment (mvollmer) [inline pkg/storaged/iscsi/session.jsx:50]: Woohoo!

**Review by mvollmer — DISMISSED** [review]
Thanks! The list of ingored errors is long and pretty arbitrary, but this might catch some regressions and then we win.

**Comment by jelly** [comment]
> Thanks! The list of ingored errors is long and pretty arbitrary, but this might catch some regressions and then we win.

Cleaning that up is a huge effort and not always possible. As requested on Matrix here is the statistics for the type of errors:

```
[('TS7006', 4191),
 ('TS2339', 1773),
 ('TS7005', 778),
 ('TS7031', 768),
 ('TS7053', 489),
 ('TS2345', 318),
 ('TS18046', 293),
 ('TS2322', 276),
 ('TS7034', 276),
 ('TS2739', 192),
 ('TS2531', 181),
 ('TS18047', 119),
 ('TS2741', 76),
 ('TS2769', 66),
 ('TS2554', 58),
 ('TS2683', 42),
 ('TS7008', 36),
 ('TS2353', 35),
 ('TS18048', 33),
 ('TS2307', 29),
 ('TS2551', 17),
 ('TS7023', 17),
 ('TS2810', 15),
 ('TS2532', 10),
 ('TS2571', 8),
 ('TS2550', 8),
 ('TS2533', 8),
 ('TS2375', 7),
 ('TS2304', 6),
 ('TS7016', 6),
 ('TS2540', 5),
 ('TS7019', 5),
 ('TS7022', 3),
 ('TS2538', 3),
 ('TS2367', 2),
 ('TS7015', 2),
 ('TS2349', 2),
 ('TS2363', 2),
 ('TS2447', 2),
 ('TS2407', 2),
 ('TS2305', 2),
 ('TS2365', 2),
 ('TS2559', 2),
 ('TS7024', 2),
 ('TS2488', 1),
 ('TS2693', 1),
 ('TS1345', 1),
 ('TS2581', 1),
 ('TS7009', 1),
 ('TS2790', 1),
 ('TS2362', 1),
 ('TS2740', 1),
 ('TS2464', 1)]
 ```

**Review by mvollmer — APPROVED** [review]
Thanks! Now I want a mode that outputs the statistics! :-)

**Pull request merged**